runtime/src/qio/qio_formatted.c

```diff
@@ -361,6 +361,84 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
   return 0;
 }
 
+qioerr read_entire_file(const int threadsafe,qio_channel_t* restrict ch, const char* restrict* restrict out, int64_t* restrict len_out, ssize_t maxlen)
+{  
+  qioerr err,err_reading_file=0;
+  int64_t file_length;
+  char* ret=NULL;
+  ssize_t amt=0;
+  err_t errcode;
+  int64_t fileLenGuess=1024;
+  ssize_t total_amt_read=0;
+  bool err_flag;
+  if( threadsafe ) 
+  {
+    err = qio_lock(&ch->lock);
+    if( err ) return err;
+  }
+
+  err=qio_file_length(ch->file,&file_length);   // obtain file length
+ 
+  if(!err)                                      // no error in obtaining file length
+  {
+   
+    ret=qio_malloc(file_length+1);              //allocate initial buffer
+    ret[0]='\0';
+    err=qio_channel_read(false, ch, ret,file_length+1, &amt);    //file_length+1 instead of file_length -->I try to read one more character than the prescribed 
+                                                                 //file length . That way, if we have already reached the file end, reading 1 more character will 
+                                                                 // will generate EOF. But, if the file length has increased while reading, we have just read one more
+                                                                 // character.
+    if(qio_err_to_int(err)==EEOF)                                // Here,checking whether file end is reached.
+    {
+       err=0;
```

Comment:
If the file became shorter between the qio_file_length and the qio_channel_read call then we will have allocated more than necessary and could realloc here.

Reply:
(I am not sure that case matters, though.)

Reply:
@mppf In my opinion, I do not feel this case is valid. Through programs, we can read and write a file, deletion of file contents does not happen in the truest sense, we basically overwrite file contents, so the entire contents of the previous version becomes invalid. So,that way,if we speak of file shortening, we should be rereading the entire file instead of reallocating to prevent wastage of memory.

Reply:
I don't really believe that argument but I don't think you need to handle the case because allocating extra memory in that rare case seems like an issue we don't have to solve here.